File: go/pkg/protocol/connection.go
package protocol

import "time"

type CreateConnectionRequest struct {
	Name          string            `json:"name" msgpack:"name"`
	Kind          DatabaseKind      `json:"kind" msgpack:"kind"`
	Host          string            `json:"host,omitempty" msgpack:"host,omitempty"`
	Port          int               `json:"port,omitempty" msgpack:"port,omitempty"`
	Database      string            `json:"database,omitempty" msgpack:"database,omitempty"`
	Username      string            `json:"username,omitempty" msgpack:"username,omitempty"`
	PasswordRef   string            `json:"password_ref,omitempty" msgpack:"password_ref,omitempty"`
	Password      string            `json:"password,omitempty" msgpack:"password,omitempty"`
	Options       map[string]string `json:"options,omitempty" msgpack:"options,omitempty"`
	SSHTunnel     *SshTunnelConfig  `json:"ssh_tunnel,omitempty" msgpack:"ssh_tunnel,omitempty"`
	DefaultSchema string            `json:"default_schema,omitempty" msgpack:"default_schema,omitempty"`
	// ReadOnly 为 true 时 app-core 拒绝写语句与 KV 写（behavior.md §12.3）。
	// 用指针保证缺省即 false 且线上不编码（与 Rust Option<bool> 对齐）。
	ReadOnly *bool `json:"read_only,omitempty" msgpack:"read_only,omitempty"`
}

type UpdateConnectionRequest struct {
	Name          *string           `json:"name,omitempty" msgpack:"name,omitempty"`
	Host          *string           `json:"host,omitempty" msgpack:"host,omitempty"`
	Port          *int              `json:"port,omitempty" msgpack:"port,omitempty"`
	Database      *string           `json:"database,omitempty" msgpack:"database,omitempty"`
	Username      *string           `json:"username,omitempty" msgpack:"username,omitempty"`
	PasswordRef   *string           `json:"password_ref,omitempty" msgpack:"password_ref,omitempty"`
	Password      string            `json:"password,omitempty" msgpack:"password,omitempty"`
	Options       map[string]string `json:"options,omitempty" msgpack:"options,omitempty"`
	SSHTunnel     *SshTunnelConfig  `json:"ssh_tunnel,omitempty" msgpack:"ssh_tunnel,omitempty"`
	DefaultSchema *string           `json:"default_schema,omitempty" msgpack:"default_schema,omitempty"`
	ReadOnly      *bool             `json:"read_only,omitempty" msgpack:"read_only,omitempty"`
}

type ConnectionInfo struct {
	ID            string            `json:"id" msgpack:"id"`
	Name          string            `json:"name" msgpack:"name"`
	Kind          DatabaseKind      `json:"kind" msgpack:"kind"`
	Host          string            `json:"host,omitempty" msgpack:"host,omitempty"`
	Port          int               `json:"port,omitempty" msgpack:"port,omitempty"`
	Database      string            `json:"database,omitempty" msgpack:"database,omitempty"`
	Username      string            `json:"username,omitempty" msgpack:"username,omitempty"`
	Options       map[string]string `json:"options,omitempty" msgpack:"options,omitempty"`
	SSHTunnel     *SshTunnelConfig  `json:"ssh_tunnel,omitempty" msgpack:"ssh_tunnel,omitempty"`
	DefaultSchema string            `json:"default_schema,omitempty" msgpack:"default_schema,omitempty"`
	ReadOnly      *bool             `json:"read_only,omitempty" msgpack:"read_only,omitempty"`
	CreatedAt     time.Time         `json:"created_at" msgpack:"created_at"`
	UpdatedAt     time.Time         `json:"updated_at" msgpack:"updated_at"`
}

type ConnectionStatus struct {
	ID            string  `json:"id" msgpack:"id"`
	Connected     bool    `json:"connected" msgpack:"connected"`
	ServerVersion string  `json:"server_version,omitempty" msgpack:"server_version,omitempty"`
	LatencyMs     float64 `json:"latency_ms,omitempty" msgpack:"latency_ms,omitempty"`
	Error         string  `json:"error,omitempty" msgpack:"error,omitempty"`
}

File: rust/crates/protocol/src/connection.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::common::{ConnectionId, DatabaseKind, SshTunnelConfig};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateConnectionRequest {
    pub name: String,
    pub kind: DatabaseKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub host: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub database: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub password_ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub password: Option<String>,
    #[serde(default, skip_serializing_if = "std::collections::HashMap::is_empty")]
    pub options: std::collections::HashMap<String, String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ssh_tunnel: Option<SshTunnelConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_schema: Option<String>,
    /// 只读连接（behavior.md §12.3）：为 true 时 app-core 拒绝写语句与 KV 写。
    #[serde(skip_serializing_if = "Option::is_none")]
    pub read_only: Option<bool>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct UpdateConnectionRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub host: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub database: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub password_ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub password: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<std::collections::HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ssh_tunnel: Option<SshTunnelConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_schema: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub read_only: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionInfo {
    pub id: ConnectionId,
    pub name: String,
    pub kind: DatabaseKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub host: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub database: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(default, skip_serializing_if = "std::collections::HashMap::is_empty")]
    pub options: std::collections::HashMap<String, String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ssh_tunnel: Option<SshTunnelConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_schema: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub read_only: Option<bool>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionStatus {
    pub id: ConnectionId,
    pub connected: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub server_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub latency_ms: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}
